src/pybamm/meshes/scikit_fem_submeshes_3d.py

```diff
@@ -0,0 +1,377 @@
+                    np.linalg.det(
+                        np.column_stack([v[1] - v[0], v[2] - v[0], v[3] - v[0]])
+                    )
+                )
+                / 6
+            )
+            volumes.append(vol)
+        tetrahedra = tetrahedra[np.array(volumes) > 1e-10]
+        mesh = skfem.MeshTet(nodes_3d.T, np.array(tetrahedra).T)
+
+        bottom_nodes = np.arange(n_nodes_per_layer)
+        top_nodes = np.arange(nodes_3d.shape[0] - n_nodes_per_layer, nodes_3d.shape[0])
+
+        outer_radius_nodes_2d = np.where(
+            np.isclose(np.linalg.norm(points_2d, axis=1), radius, rtol=1e-3)
+        )[0]
+        side_nodes = np.concatenate(
+            [outer_radius_nodes_2d + (i * n_nodes_per_layer) for i in range(n_z)]
+        )
+
+        boundary_facets = mesh.boundary_facets()
+        facet_nodes = mesh.facets[:, boundary_facets]
+
+        bottom_facets = []
+        top_facets = []
+        side_facets = []
+
+        for i, facet_idx in enumerate(boundary_facets):
+            nodes_in_facet = facet_nodes[:, i]
+
+            if np.all(np.isin(nodes_in_facet, bottom_nodes)):
+                bottom_facets.append(facet_idx)
+            elif np.all(np.isin(nodes_in_facet, top_nodes)):
+                top_facets.append(facet_idx)
+            elif np.any(np.isin(nodes_in_facet, side_nodes)):
+                side_facets.append(facet_idx)
+
+        boundaries = {}
+        if bottom_facets:
+            boundaries["bottom cap"] = np.array(bottom_facets)
+        if top_facets:
+            boundaries["top cap"] = np.array(top_facets)
+        if side_facets:
+            boundaries["side wall"] = np.array(side_facets)
+
+        if boundaries:
+            all_boundary_nodes = set()
+            for facet_list in boundaries.values():
+                all_boundary_nodes.update(mesh.facets[:, facet_list].flatten())
+            mesh = laplacian_smooth(mesh, list(all_boundary_nodes), iterations=2)
+
+        return mesh.with_boundaries(boundaries)
+
+    def __call__(self, lims, npts):
+        """
+        Main entry point called by PyBaMM's Discretisation.
+
+        Parameters
+        ----------
+        lims : dict
+            Dictionary of domain limits
+        npts : dict
+            Dictionary of target number of points
+
+        Returns
+        -------
+        ScikitFemSubMesh3D
+            Generated 3D submesh
+        """
+        h = self.gen_params.get("h", 0.3)
+
+        if self.geom_type == "box":
+            x_key = next(k for k in lims if k.name == "x")
+            y_key = next(k for k in lims if k.name == "y")
+            z_key = next(k for k in lims if k.name == "z")
+            x_lim = tuple(lims[x_key].values())
+            y_lim = tuple(lims[y_key].values())
+            z_lim = tuple(lims[z_key].values())
+            mesh = self._make_box_mesh(x_lim, y_lim, z_lim, h)
+
+        elif self.geom_type == "cylinder":
+            radius = self.gen_params.get("radius", 0.4)
```

Comment:
I don't think these should have defaults. It might lead to incorrect geometries being created silently by accident.